rfcs/013 - Datastore Revisited RFC.md

```diff
@@ -0,0 +1,137 @@
+
+## Proposal
+
+The current issue with `@std/datastore` is that it is an unintentional source of unsafe casting between types. Eg
+
+```ln
+from @std/app import start, print, exit
+from @std/datastore import namespace, set, getOr
+
+on start {
+  const foo = namespace('foo');
+  foo.set('bar', 5);
+  print(foo.getOr('bar', 1));
+  print(foo.getOr('bar', 1.0));
+  emit exit 0;
+}
+```
+
+not only works, but it produces the following output
+
+```
+5
+0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000025
+```
+
+The reason why this is a bug and not a feature is that it's an *unsafe* casting between types, so this
+
+```ln
+from @std/app import start, print, exit
+from @std/datastore import namespace, set, getOr
+
+on start {
+  const foo = namespace('foo');
+  foo.set('bar', 5);
+  print(foo.getOr('bar', 1));
+  print(foo.getOr('bar', ''));
+  emit exit 0;
+}
+```
+
+produces this
+
+```
+thread 'tokio-runtime-worker' panicked at 'range end index 5 out of range for slice of length 0', src/vm/memory.rs:628:29
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: JoinError::Panic(...)', src/vm/event.rs:260:14
+```
+
+a hard crash of the runtime as it tries to access a number as if it was a string. With a stated goal of eliminating as many runtime errors as possible so you can be sure that your backend is going to continue running, having this unsafe casting footgun was absolutely not the intention of the `datastore` library.
+
+There are multiple ways to resolve this such that the actual datastore opcodes are left alone while the compiler guarantees that their usage is safe. I'm not actually sure which one(s) I like best, so I'm listing them all in here and we can debate the pros/cons.
+
+### Minimally-invasive with restored type introspection
+
+We recently removed the ability to get the type of a variable as a string with the `type variableName` syntax, but if we restore that, it should be possible to force type safety by combining the type string with either the namespace or the key string, such that trying to get the same key name with a default value of a different type will find that it doesn't exist and return the provided default.
+
+This could provide zero actual API call changes except to the `del` function, which would need to be given an example of the type of data to delete in order to find it. This has the smallest change requirement, though it does require bringing back the string-based type manipulation code.
+
+### Generic Functions (in addition to the interface functions that currently exist)
```

Comment:
I like this, it'll also remove the need for `any0`, `any1`, `any2` type variables. Something to keep in mind is how the generics would look with anonymous functions (since generics only apply to functions with params and/or a return type, we only have to worry about the `fn ([params])[: ret] { ... }` syntax here...)

```
fn foo<A>(arg1: int64, arg2: string): A { ... }
fn <A>(arg1: int64, arg2: string): A { ... }
```

Honestly, I'm not a fan of having the `<A>` right after `fn` in that scenario - my issue is with the usage of `<` and `>` without a name preceding it. It still makes sense and I'm not opposed to it, I just think if we can find a better syntax we should use it. FWIW, I do like the generics design for Go 2 (the first alternative that comes to mind), but I'm not sure how it would fit in with more Alan:
```
fn foo[A](arg1: int64, arg2: string): A { ... }
fn [A](arg1: int64, arg2: string): A { ... }
```

Reply:
WRT `fn foo<A as Orderable>(...`, I don't think that's the worst thing. In fact, I think we can make the current interface-accepting syntax just sugar for that.
```
fn foo(arg1: Orderable, arg2: int64): string { ... }
// the above gets desugared into (and is equivalent) to this:
fn foo<A as Orderable>(arg1: A, arg2: int64): string { ... }
```

This way, none of the behavior changes, we've just introduced syntax for constraining multiple values to the same type. In fact, I think unconstrained generic type variables should have an implicit `as any`
```
fn foo<A>(arg1: int64, arg2: string): A { ... }
// the above gets desugared into (and is equivalent) to this:
fn foo<A as any>(arg1: int64, arg2: string): A { ... }
```

Reply:
This approach would still need the `type variableName` syntax of the first approach, too, in order to provide that data to the `ds*` opcodes, just to be clear. Only the very last option where we dig into redoing those opcodes would that potentially not be the case.